### PR TITLE
[Bugfix][SpecDecode] Adjust Eagle model architecture to align with intended design

### PR DESCRIPTION
**Pull request by llsj14**
## Summary
#9565
#11126

1. **disable norm**: Removed input layer normalization and output normalization.
2. **add residual**: Added a residual path at the end of the Llama model. This was omitted because output normalization was disabled in the first step.

## Experiment
Below are the experimental results from the above trials.

- model: `Llama-2-7b-chat-hf / EAGLE-llama2-chat-7B` with `K=1`
- dataset: MT-Bench
- input/output length: 128/128
- sampling setting:
  - multinomial: top_k=-1, top_p=1.0, temp=1.0
  - greedy: top_k=1, top_p=1.0, temp=1.0

| Approach | Accept Rate (Multinomial) | Accept Rate (Greedy) |
| -------- | ------------------------ | -------------------- |
| as-is | 0.131 | 0.308 |
| 1 (disable norm) | 0.391 | 0.529 |
| 1+2 (disable norm + add residual) | 0.565 | 0.619 |

## Additional Experiment
- model: `Llama-2-7b-chat-hf / EAGLE-llama2-chat-7B` with different `K`
- dataset: c4 / MT-Bench
- input/output length: 1024/128(c4), 128/128(MT-Bench)
- sampling setting:
  - multinomial: top_k=-1, top_p=1.0, temp=1.0
  - greedy: top_k=1, top_p=1.0, temp=1.0
### Llama-2-7b-chat-hf / EAGLE-llama2-chat-7B
  |   | c4 |   | mt-bench |  
-- | -- | -- | -- | -- | --
  | avg. accept rate | Multinomial | Greedy | Multinomial | Greedy
K=1 | as-is | 0.103 | 0.302 | 0.136 | 0.302
  | to-be | 0.393 | 0.508 | 0.58 | 0.607
K=2 | as-is | 0.101 | 0.209 | 0.112 | 0.229
  | to-be | 0.329 | 0.416 | 0.486 | 0.543
K=3 | as-is | 0.111 | 0.167 | 0.117 | 0.192
  | to-be | 0.306 | 0.36 | 0.437 | 0.495

### Meta-Llama-3-8B-Instruct / EAGLE-LLaMA3-Instruct-8B
  |   | c4 |   | mt-bench |  
-- | -- | -- | -- | -- | --
  | avg. accept rate | Multinomial | Greedy | Multinomial | Greedy
K=1 | as-is | 0.115 | 0.441 | 0.156 | 0.561
  | to-be | 0.412 | 0.473 | 0.56 | 0.593
K=2 | as-is | 0.121 | 0.286 | 0.143 | 0.376
  | to-be | 0.339 | 0.365 | 0.449 | 0.468
K=3 | as-is | 0.138 | 0.236 | 0.152 | 0.301
  | to-be | 0.304 | 0.307 | 0.399 | 0.414

### CODE REVIEW

File: vllm/model_executor/models/llama.py [hunk 1]
```diff
@@ -315,6 +315,12 @@ def __init__(self,
             )
         else:
             self.embed_tokens = PPMissingLayer()
+
+        if config.num_hidden_layers==1:
+            self.eagle = True
+        else:
+            self.eagle = False
+
```
Comment (llsj14) [inline vllm/model_executor/models/llama.py:323]: I think this part should not be added on llama model implementation. I'm planning to modify `DummyNorm` instead.

**Comment by llsj14** [comment]
@LiuXiaoxuanPKU @sroy745 Could you check this PR please?

File: vllm/model_executor/models/eagle.py [hunk 2]
```diff
@@ -17,6 +17,19 @@
 from .utils import maybe_prefix
 
 
+class DummyInputLayerNorm(nn.Module):
+
+    def forward(self, x):
+        return x
+
+
+class DummOutputNorm(nn.Module):
```
Comment (sroy745) [inline vllm/model_executor/models/eagle.py:26]: nit : Since this is doing normalization wondering if we can call it EagleOutputNorm instead of DummOutputNorm? 

If we want to retain the name consider renaming to DummyOutputNorm instead of DummOutputNorm.

**Review by sroy745 — COMMENTED** [review]
LGTM.
Thanks for the fix.
@llsj14 I am wondering if there is any easy way to find the acceptance rate of the reference github implementation for the mt-benchmark dataset?

File: vllm/model_executor/models/eagle.py [hunk 3]
```diff
@@ -46,10 +59,16 @@ def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
...
         self.model = model_cls(vllm_config=vllm_config,
                                prefix=maybe_prefix(prefix, "model"))
+
         self.fc = nn.Linear(config.model.hidden_size * 2,
                             config.model.hidden_size,
                             bias=getattr(self.config, "eagle_fc_bias", False))
 
+        # Modify layer normalization and residual connections as suggested
+        # in the EAGLE framework: https://github.com/SafeAILab/EAGLE
+        self.model.model.layers[0].input_layernorm = DummyInputLayerNorm()
```
Comment (sroy745) [inline vllm/model_executor/models/eagle.py:69]: Can you please update the comment in L25 ("In reference, LlamaDecoderLayer implementation doesn't have 
       input_layernorm for 1st decoder layer") ? I think this comment may no longer be applicable after this change?

File: vllm/model_executor/models/eagle.py (hunk 2 above)
Comment (llsj14) [reply vllm/model_executor/models/eagle.py]: Actually, in this function, it doesn't perform normalization but only applies a residual operation. Therefore, I corrected the name DummOutputNorm, which contained a typo, to DummyOutputNorm.

File: vllm/model_executor/models/eagle.py (hunk 3 above)
Comment (llsj14) [reply vllm/model_executor/models/eagle.py:70]: I updated the comment in L25 to reflect that our implementation also follows the approach of the EAGLE framework.

**Comment by llsj14** [comment]
>@llsj14 I am wondering if there is any easy way to find the acceptance rate of the reference github implementation for the mt-benchmark dataset?

I need some advice on applying the MT-Bench dataset, as I’m unsure how to properly use the prompts, especially those with `conversation_a` and `conversation_b` in the MT-Bench. For my experiment, I only used the content of `conversation_a` and truncated it to an input length of 128. I suspect this might be causing discrepancies between my results and those reported in the paper(But I also have limited information about how they used the MT-Bench dataset.).

For benchmarking, I used llmperf, and below is the code snippet showing how I processed the MT-Bench prompts:
```python
                        prompt_str = traindata[index]['conversation_a'][0]['content'] #MT-BENCH
                        #prompt_str = traindata[index]['text'] #for C4 dataset
                        encoded_token = tokenizer.encode(prompt_str)                                         
                        token_length = len(encoded_token)

                        if token_length >= mean_input_tokens:
                            index += 1
                            encoded_token = encoded_token[:mean_input_tokens]
                            prompt_str = tokenizer.decode(encoded_token)
                            prompt = (prompt_str, get_token_length(prompt_str))
                            break
```

To compare my data with the reference, my colleague checked the acceptance rate using the EAGLE implementation and found that it has a similar acceptance rate to the C4 dataset with k=1 settings.

Do you have any feedback on this situation?

File: vllm/model_executor/models/eagle.py (hunk 2 above)
Comment (llsj14) [reply vllm/model_executor/models/eagle.py]: I’m open to any comments if the name is not reasonable.

**Comment by sroy745** [comment]
> > @llsj14 I am wondering if there is any easy way to find the acceptance rate of the reference github implementation for the mt-benchmark dataset?
> 
> I need some advice on applying the MT-Bench dataset, as I’m unsure how to properly use the prompts, especially those with `conversation_a` and `conversation_b` in the MT-Bench. For my experiment, I only used the content of `conversation_a` and truncated it to an input length of 128. I suspect this might be causing discrepancies between my results and those reported in the paper(But I also have limited information about how they used the MT-Bench dataset.).
> 
> For benchmarking, I used llmperf, and below is the code snippet showing how I processed the MT-Bench prompts:
> 
> ```python
>                         prompt_str = traindata[index]['conversation_a'][0]['content'] #MT-BENCH
>                         #prompt_str = traindata[index]['text'] #for C4 dataset
>                         encoded_token = tokenizer.encode(prompt_str)                                         
>                         token_length = len(encoded_token)
> 
>                         if token_length >= mean_input_tokens:
>                             index += 1
>                             encoded_token = encoded_token[:mean_input_tokens]
>                             prompt_str = tokenizer.decode(encoded_token)
>                             prompt = (prompt_str, get_token_length(prompt_str))
>                             break
> ```
> 
> To compare my data with the reference, my colleague checked the acceptance rate using the EAGLE implementation and found that it has a similar acceptance rate to the C4 dataset with k=1 settings.
> 
> Do you have any feedback on this situation?

Thanks for the update. Good to know the acceptance rate for C4 is similar. One thing - wondering if it would be possible to run the vLLM benchmark with the ShareGPT dataset with and without the pr and see the improvement in TPOT ?

**Comment by sroy745** [comment]
Applied the ready label to run all tests while we wait for reviews from others.

**Comment by jeongin601** [comment]
> @llsj14 I am wondering if there is any easy way to find the acceptance rate of the reference github implementation for the mt-benchmark dataset?

Hi,

The EAGLE GitHub repository customizes the original MT-Bench evaluation file (available [here](https://github.com/lm-sys/FastChat/blob/main/fastchat/llm_judge/gen_model_answer.py)) to implement its evaluation process.You can find the customized MT-Bench evaluation file in the EAGLE repository [here](https://github.com/SafeAILab/EAGLE/blob/main/eagle/evaluation/gen_ea_answer_llama2chat.py). 

For the acceptance rate, the EAGLE implementation currently does not provide support for calculating the acceptance rate.

**Comment by llsj14** [comment]
>Thanks for the update. Good to know the acceptance rate for C4 is similar. One thing - wondering if it would be possible to run the vLLM benchmark with the ShareGPT dataset with and without the pr and see the improvement in TPOT ?

Sure, I’ll update the results. It might be helpful to understand the benchmark settings used by @LiuXiaoxuanPKU for the ShareGPT dataset in #9565, so you can better analyze the results.

**Comment by llsj14** [comment]
I just did my experiment with [SparseGPT dataset](https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered). The process of extracting prompts can be different from the experiment of @LiuXiaoxuanPKU in issue #9565

## Experiment settings
- model: Llama-2-7b-chat-hf / EAGLE-llama2-chat-7B
- device: A100
- sampling parameter: multinomial (top_k=-1, top_p=1.0, temperature=1.0)
- dataset: SparseGPT
- average input length: 318 tokens
- average output length: 128 tokens
- the number of requests: 512

## The part to extract prompts                                                               
```python
traindata = load_dataset("json",
data_files="https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split_no_imsorry.json",
split='train')

conversations = traindata[index].get('conversations', []) # sharegpt
if conversations:
    prompt_str = conversations[0].get('value', 'default_value') # sharegpt
else:
    index += 1
    continue
```

## Experiment results
- The results showed improvements in acceptance rate and speed-up metrics compared to the results w/o this PR.
- However, they all shower performance than vanilla case which doesn't utilize speculative decoding.

  |   | accept rate | system efficiency | inter token latency(mean) [s] | total elapsed time [s] | speed up (vs vanilla) | speed up (vs as-is)
-- | -- | -- | -- | -- | -- | -- | --
  | vanilla | - | - | 0.02 | 53.64 | - | -
K=1 | as-is | 0.112 | 0.556 | 0.036 | 81.83 | 0.66 | -
  | to-be | 0.56 | 0.78 | 0.029 | 66.58 | 0.81 | 1.23
K=2 | as-is | 0.106 | 0.374 | 0.043 | 95.59 | 0.56 | -
  | to-be | 0.476 | 0.602 | 0.03 | 68.45 | 0.78 | 1.40
K=3 | as-is | 0.115 | 0.28 | 0.047 | 107.01 | 0.50 | -
  | to-be | 0.43 | 0.477 | 0.033 | 74.35 | 0.72 | 1.44

**Comment by Lin-Qingyang-Alec** [comment]
Maybe you could use nn.Identity replace of DummyInputLayerNorm

**Comment by llsj14** [comment]
>Maybe you could use nn.Identity replace of DummyInputLayerNorm

Thank you for the recommendation, @Lin-Qingyang-Alec.

I didn’t know about nn.Identity, which behaves exactly like DummyInputLayerNorm in my code. However, I think it’s not a bad idea to use both DummyInputLayerNorm and DummyOutputNorm. I was looking for an equivalent implementation to DummyOutputNorm that performs residual operations, but I couldn’t find a similar one in the nn library.
If I need to change it, I’d prefer to modify both at the same time.

**Comment by LiuXiaoxuanPKU** [comment]
Hi folks, a high level question, please correct me if I am wrong. It seems EAGLE changes the baseline model (removing the layernorm of the first decode layer). Does it mean EAGLE might change the model accuracy compared to the base model?

**Comment by llsj14** [comment]
@LiuXiaoxuanPKU 
Hi,
ah input layernorm is only removed for the EAGLE model, not the baseline model, so I believe this change doesn't affect model accuracy.
I modified the EAGLE model to align with the EAGLE framework implementation:

- Input LayerNorm: [Reference Code](https://github.com/SafeAILab/EAGLE/blob/3653144d76c56d763f3a11d6984841be3d501fcc/eagle/modeling_eagle.py#L532)
- Residual Connection: [Reference Code](https://github.com/SafeAILab/EAGLE/blob/3653144d76c56d763f3a11d6984841be3d501fcc/eagle/modeling_eagle.py#L549)

**Comment by llsj14** [comment]
> If so, has anyone compared the accuracy of EAGLE and the original model on some datasets? (no numbers from the paper at least)

The accuracy can be changed with the problem of tree implementation of EAGLE, which does selection instead of sampling.
If we don't use tree, accuracy will not be changed.

File: vllm/model_executor/models/eagle.py [hunk 4]
```diff
@@ -17,14 +17,28 @@
...
+class DummyInputLayerNorm(nn.Module):
+
+    def forward(self, x):
+        return x
+
+
+class DummyOutputNorm(nn.Module):
+
+    def forward(self, x, residual):
+        x = x + residual
```
Comment (LiuXiaoxuanPKU) [inline vllm/model_executor/models/eagle.py:29]: Do we need to handle cases when residual might be None?
such as [here](https://github.com/vllm-project/vllm/blob/973f5dc581c35a9c5b9176116e2f42f3f97d0d01/vllm/model_executor/layers/layernorm.py#L45).

**Review by LiuXiaoxuanPKU — APPROVED** [review]
Thanks for the fix and all the clarification, LGTM and only a minor comment.

File: vllm/model_executor/models/eagle.py (hunk 4 above)
Comment (llsj14) [reply vllm/model_executor/models/eagle.py]: yeah, I think it is needed. I added condition checking whether `residual` is None or not.

**Comment by skylee-01** [comment]
What is the setting for speculate_num used in the experiment?

**Comment by llsj14** [comment]
@skylee-01 
For tables with accept rate, K represents to the number of speculative tokens.
In the output token distribution test, the number of speculative tokens was set to 3.

**Comment by LiuXiaoxuanPKU** [comment]
Sorry for the delay here, can we mimic the behavior as [this](https://github.com/vllm-project/vllm/blob/973f5dc581c35a9c5b9176116e2f42f3f97d0d01/vllm/model_executor/layers/layernorm.py#L70C1-L73C31). After this change, the PR is good to go! @llsj14

**Comment by DarkLight1337** [comment]
Looks like your PR broke speculative decoding tests:

https://buildkite.com/vllm/ci/builds/11840#019453bc-9cd9-42db-9c6f-545891b7689e/6-12007

Can you fix it?

**Comment by llsj14** [comment]
@DarkLight1337 
Oh, okay. I'll check it out. However, I'm curious why this test is failing, as the PR didn’t show any failed test results.

**Comment by DarkLight1337** [comment]
I guess the tests were skipped in the CI because none of the listed dependencies were changed. Can you update the dependencies in the test pipeline YAML to prevent similar issues in the future?

**Comment by DarkLight1337** [comment]
I guess the tests were skipped in the CI because none of the listed dependencies were changed. Can you update the dependencies in the test pipeline YAML to prevent similar issues in the future?

**Comment by llsj14** [comment]
>I guess the tests were skipped in the CI because none of the listed dependencies were changed. Can you update the dependencies in the test pipeline YAML to prevent similar issues in the future?

yeah, I will. Can you guide me to the exact file location? I'm not familiar with this.

**Comment by DarkLight1337** [comment]
It is under `.buildkite` directory
